runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -722,9 +741,72 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
             TR::Node *replacement = acmpHandler(this, node);
             TR_ASSERT_FATAL_WITH_NODE(node, replacement == node, "can't replace n%un here",
                   node->getGlobalIndex());
+
+            transformObjectCmp = true;
             }
          }
-      else
+      else if ((isLhsValue == TR_yes) && (isRhsValue == TR_yes) && (lhsClass == rhsClass) && lhsClass)
+         {
+         const  TR::TypeLayout *fieldTypeLayout = comp()->typeLayout(lhsClass);
+         size_t fieldCount = fieldTypeLayout->count();
+
+         if (fieldCount == 1)
+            {
+            const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(0);
+            TR::DataType dataType = fieldEntry._datatype;
+
+            if (dataType.isIntegral() || dataType.isDouble() || dataType.isFloat())
+               {
+               transformObjectCmp = true;
+
+               _valueTypesHelperCallsToBeFolded.add(
+                  new (trStackMemory()) ObjectComparisonHelperCallTransform(_curTree, node,
+                     ValueTypesHelperCallTransform::InsertDebugCounter | ValueTypesHelperCallTransform::IsVTSingleFieldCmp, lhsClass));
+
+               if (trace())
+                  {
+                  traceMsg(comp(), "%s: Add delayed transformation for callNode n%dn: lhsNode n%dn rhsNode n%dn fieldCount 1\n", __FUNCTION__,
+                     node->getGlobalIndex(), lhsNode->getGlobalIndex(), rhsNode->getGlobalIndex());
+                  }
+               }
+            }
+         else if (isObjectInequalityCompare && comp()->cg()->getSupportsArrayCmp())
+            {
+            bool inlineMultiFieldsVTCmp = true;
+            int32_t totalFieldSize = 0;
+
+            for (size_t idx = 0; idx < fieldCount; idx++)
+               {
+               const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(idx);
+               TR::DataType dataType = fieldEntry._datatype;
+
+               if (!dataType.isIntegral())
```

Comment:
It looks like I was mistaken about using `arraycmp` for `float` and `double` fields.  It can't be used in the presence of such fields, as all NaN values are folded to the same bit representations by `Float.float.ToIntBits` and `Double.doubleTtoLongBits`.  So those fields cannot be compared with a direct comparison or using `arraycmp`.  

As Tobi mentioned in a side conversation, you could adjust this simply to check `J9ClassCanSupportFastSubstitutability` on the class.

Reply:
Addressed in afbf4a9